File: server/src/main/java/io/microconfig/server/rest/ConfigOptionsResolver.java
package io.microconfig.server.rest;

import io.microconfig.server.configs.ConfigOptions;
import lombok.RequiredArgsConstructor;
import org.springframework.core.MethodParameter;
import org.springframework.stereotype.Service;
import org.springframework.web.bind.support.WebDataBinderFactory;
import org.springframework.web.context.request.NativeWebRequest;
import org.springframework.web.method.support.HandlerMethodArgumentResolver;
import org.springframework.web.method.support.ModelAndViewContainer;

import java.util.Map;

import static java.util.Arrays.stream;
import static java.util.Collections.emptyMap;
import static java.util.stream.Collectors.toMap;

@Service
@RequiredArgsConstructor
public class ConfigOptionsResolver implements HandlerMethodArgumentResolver {

    @Override
    public boolean supportsParameter(MethodParameter parameter) {
        return ConfigOptions.class == parameter.getParameterType();
    }

    @Override
    public ConfigOptions resolveArgument(MethodParameter _1,
                                         ModelAndViewContainer _2,
                                         NativeWebRequest webRequest,
                                         WebDataBinderFactory _3) {
        var branch = webRequest.getHeader("X-BRANCH");
        var tag = webRequest.getHeader("X-TAG");
        var vars = vars(webRequest);
        return new ConfigOptions(branch, tag, vars);
    }

    private Map<String, String> vars(NativeWebRequest webRequest) {
        var headerValues = webRequest.getHeaderValues("X-VAR");
        if (headerValues == null) return emptyMap();
        return stream(headerValues)
                .map(ConfigOptionsResolver::splitVar)
                .collect(toMap(s -> s[0], s -> s[1]));
    }

    private static String[] splitVar(String str) {
        int idx = str.indexOf('=');
        return new String[]{str.substring(0, idx), str.substring(idx + 1)};
    }
}

File: server/src/main/kotlin/io/microconfig/server/api/ConfigOptionsResolver.kt
package io.microconfig.server.api

import io.microconfig.server.configs.ConfigOptions
import org.springframework.core.MethodParameter
import org.springframework.stereotype.Service
import org.springframework.web.bind.support.WebDataBinderFactory
import org.springframework.web.context.request.NativeWebRequest
import org.springframework.web.context.request.ServletWebRequest
import org.springframework.web.method.support.HandlerMethodArgumentResolver
import org.springframework.web.method.support.ModelAndViewContainer

@Service
class ConfigOptionsResolver : HandlerMethodArgumentResolver {
    override fun supportsParameter(parameter: MethodParameter): Boolean {
        return ConfigOptions::class.java == parameter.parameterType
    }

    override fun resolveArgument(
        _1: MethodParameter,
        _2: ModelAndViewContainer?,
        webRequest: NativeWebRequest,
        _3: WebDataBinderFactory?
    ): ConfigOptions {
        val query = query(webRequest)
        val type = query["type"] ?: webRequest.getHeader("X-TYPE")
        val ref = query["ref"] ?: webRequest.getHeader("X-REF")
        val vars = vars(webRequest)
        return ConfigOptions(type, ref, vars)
    }

    private fun vars(webRequest: NativeWebRequest): Map<String, String> {
        val headerValues = webRequest.getHeaderValues("X-VAR") ?: return emptyMap()
        return headerValues
            .map(this::splitVar)
            .toMap()
    }

    private fun splitVar(str: String): Pair<String, String> {
        val idx = str.indexOf('=')
        val key = str.substring(0, idx)
        val value = str.substring(idx + 1)
        return Pair(key, value)
    }

    private fun query(webRequest: NativeWebRequest): Map<String, String> {
        val servletRequest = webRequest as? ServletWebRequest ?: return emptyMap()
        val query = servletRequest.request.queryString ?: return emptyMap()
        return query
            .split('&')
            .map { it.split('=') }
            .associate { it[0] to it[1] }
    }
}

